src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs

```diff
@@ -113,16 +117,55 @@ public async IAsyncEnumerable<FeatureDefinition> GetAllFeatureDefinitionsAsync()
             // Iterate over all features registered in the system at initial invocation time
             foreach (IConfigurationSection featureSection in GetFeatureDefinitionSections())
             {
+                string featureName = GetFeatureName(featureSection);
+
+                if (string.IsNullOrEmpty(featureName))
+                {
+                    continue;
+                }
+
                 //
                 // Underlying IConfigurationSection data is dynamic so latest feature definitions are returned
-                yield return  _definitions.GetOrAdd(featureSection.Key, (_) => ReadFeatureDefinition(featureSection));
+                yield return  _definitions.GetOrAdd(featureName, (_) => ReadFeatureDefinition(featureSection));
+            }
+        }
+
+        private void EnsureInit()
+        {
+            if (_initialized == 0)
+            {
+                IConfiguration featureManagementConfigurationSection = _configuration
+                    .GetChildren()
+                    .FirstOrDefault(section =>
+                        string.Equals(
+                            section.Key,
+                            ConfigurationFields.FeatureManagementSectionName,
+                            StringComparison.OrdinalIgnoreCase));
+
+                if (featureManagementConfigurationSection == null && RootConfigurationFallbackEnabled)
+                {
+                    featureManagementConfigurationSection = _configuration;
+                }
+
+                bool hasAzureAppConfigurationFeatureFlagSchema = featureManagementConfigurationSection != null && 
+                    HasAzureAppConfigurationFeatureFlagSchema(featureManagementConfigurationSection);
+
+                lock (_lock)
+                {
+                    if (Interlocked.Read(ref _initialized) == 0)
```

Comment:
lock can sure only one thread can enter this block, what is the reason for using `Interlocked.Read(ref _initialized)` to atomically get the value?

Mark int32 type `_initialized` as `volatile` then judge `_initialized == 0` is fine. Am I missing anything?

Reply:
Or use ` if (Interlocked.CompareExchange(ref _initialized, 1, 0))==0 {_azureAppConfigurationFeatureFlagSchemaEnabled = hasAzureAppConfigurationFeatureFlagSchema; }` is also fine.

Reply:
Interlocked avoids cached read from CPU. Marking the variable as volatile will have lasting effect throughout the lifetime of the application whereas interlocked usage in the init check only happens during initialization.